cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/server.go

```diff
@@ -81,3 +80,132 @@ func (s *Server) GetPackageRepositories(ctx context.Context, request *corev1.Get
 		Repositories: responseRepos,
 	}, nil
 }
+
+// GetAvailablePackages streams the available packages based on the request.
+func (s *Server) GetAvailablePackages(ctx context.Context, request *corev1.GetAvailablePackagesRequest) (*corev1.GetAvailablePackagesResponse, error) {
+	log.Infof("+GetAvailablePackages(namespace=[%s])", request.Namespace)
+	repos, err := getHelmRepos(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	responsePackages := []*corev1.AvailablePackage{}
+	for _, repoUnstructured := range repos.Items {
+		name, found, err := unstructured.NestedString(repoUnstructured.Object, "metadata", "name")
+		if err != nil || !found {
+			log.Errorf("required field metadata.name not found on HelmRepository: %w:\n%v", err, repoUnstructured.Object)
+			continue
+		}
+
+		// see docs at https://fluxcd.io/docs/components/source/helmrepositories/
+		conditions, found, err := unstructured.NestedSlice(repoUnstructured.Object, "status", "conditions")
+		if err != nil || !found {
+			log.Infof("Skipping packages for repository [%s] because it has not reached 'Ready' state:%w\n%v", name, err, repoUnstructured.Object)
+			continue
+		}
+
+		ready := false
+		for _, conditionUnstructured := range conditions {
+			if conditionAsMap, ok := conditionUnstructured.(map[string]interface{}); ok {
+				if typeString, ok := conditionAsMap["type"]; ok && typeString == "Ready" {
+					if statusString, ok := conditionAsMap["status"]; ok && statusString == "True" {
+						if reasonString, ok := conditionAsMap["reason"]; ok && reasonString == "IndexationSucceed" {
+							ready = true
+							break
+						}
+					}
+				}
+			}
+		}
+
+		if !ready {
+			log.Infof("Skipping packages for repository [%s] because it has not reached 'Ready' state:n%v", name, repoUnstructured.Object)
+		} else {
+			url, found, err := unstructured.NestedString(repoUnstructured.Object, "status", "url")
+			if err != nil || !found {
+				log.Infof("field status.url not found on HelmRepository: %w:\n%v", err, repoUnstructured.Object)
+				continue
+			}
+
+			log.Infof("Found repository: [%s], index url: [%s]", name, url)
+			repoPackages, err := readPackagesFromRepoIndex(url)
+			if err != nil {
+				// just skip this repo
+				log.Errorf("Failed to read packages for repository [%s] due to %v", name, err)
+			} else {
+				responsePackages = append(responsePackages, repoPackages...)
+			}
+		}
+	}
+	return &corev1.GetAvailablePackagesResponse{
+		Packages: responsePackages,
+	}, nil
+}
+
+func getHelmRepos(ctx context.Context) (*unstructured.UnstructuredList, error) {
```

Comment:
I was wondering why you don't have this function returning the structured type (ie. `[]core.packages.v1alpha1.PackageRepository`) but of course, you actually need flux-specific data out of here from the status, which won't be in our `PackageRepository` message.

Perhaps it'll prove worthwhile having a `PackageRepository.localIndexYAML` if that becomes a pattern, let's see.

Reply:
sure